src/services/codefixes/inferFromUsage.ts

```diff
@@ -719,10 +751,11 @@ namespace ts.codefix {
                     stringIndexInfo = checker.createIndexInfo(recur(usageContext.stringIndexContext), /*isReadonly*/ false);
                 }
 
-                return checker.createAnonymousType(/*symbol*/ undefined!, members, callSignatures, constructSignatures, stringIndexInfo, /*numberIndexInfo*/ undefined); // TODO: GH#18217
-            }
-            else {
-                return undefined;
+                const structuralInference = checker.createAnonymousType(/*symbol*/ undefined!, members, callSignatures, constructSignatures, stringIndexInfo, /*numberIndexInfo*/ undefined); // TODO: GH#18217
+                if (previous && checker.isEmptyObjectType(previous)) {
+                    return checker.mapType(previous, t => checker.isEmptyObjectType(t) ? structuralInference : t);

```

Comment:
I had to stare at this for a while until I realized `isEmptyObjectTypes` returns true non-empty non-object types like `42 | {}`. And `mapType` has nothing to do with mapped types, why on earth would I think that?